serde-generate/src/typescript.rs

```diff
@@ -0,0 +1,857 @@
+
+        write!(
+            self.out,
+            "static deserialize{}(deserializer: Deserializer): {} {{",
+            name,
+            self.quote_type(format0),
+        )?;
+        self.out.indent();
+        match format0 {
+            Option(format) => {
+                write!(
+                    self.out,
+                    r#"
+const tag = deserializer.deserializeOptionTag();
+if (!tag) {{
+    return null;
+}} else {{
+    return {};
+}}
+"#,
+                    self.quote_deserialize(format),
+                )?;
+            }
+
+            Seq(format) => {
+                write!(
+                    self.out,
+                    r#"
+const length = deserializer.deserializeLen();
+const list: {} = [];
+for (let i = 0; i < length; i++) {{
+    list.push({});
+}}
+return list;
+"#,
+                    self.quote_type(format0),
+                    self.quote_deserialize(format)
+                )?;
+            }
+
+            Map { key, value } => {
+                write!(
+                    self.out,
+                    r#"
+const length = deserializer.deserializeLen();
+const obj = new Map<{0}, {1}>();
+let previousKeyStart = 0;
+let previousKeyEnd = 0;
+for (let i = 0; i < length; i++) {{
+    const keyStart = deserializer.getBufferOffset();
+    const key = {2};
+    const keyEnd = deserializer.getBufferOffset();
+    if (i > 0) {{
+        deserializer.checkThatKeySlicesAreIncreasing(
+            [previousKeyStart, previousKeyEnd],
+            [keyStart, keyEnd]);
+    }}
+    previousKeyStart = keyStart;
+    previousKeyEnd = keyEnd;
+    const value = {3};
+    obj.set(key, value);
+}}
+return obj;
+"#,
+                    self.quote_type(key),
+                    self.quote_type(value),
+                    self.quote_deserialize(key),
+                    self.quote_deserialize(value),
+                )?;
+            }
+
+            Tuple(formats) => {
+                write!(
+                    self.out,
+                    r#"
+return [{}
+];
+"#,
+                    formats
+                        .iter()
+                        .map(|f| format!("\n    {}", self.quote_deserialize(f)))
+                        .collect::<Vec<_>>()
+                        .join(",")
+                )?;
+            }
+
+            TupleArray { content, size } => {
+                write!(
+                    self.out,
+                    r#"
+const list: {} = [];
+for (let i = 0; i < {}; i++) {{
+    list.push({});
+}}
+return list;
+"#,
+                    self.quote_type(format0),
+                    size,
+                    self.quote_deserialize(content)
+                )?;
+            }
+
+            _ => panic!("unexpected case"),
+        }
+        self.out.unindent();
+        writeln!(self.out, "}}\n")
+    }
+
+    /// Make code more readable by using typedefs for complex types (e.g. instead of string | null as a type, we define OptionalString = string | null)
```

Comment:
I'm not sure I understand the rational here. TypeScript supports generic type aliases. Shouldn't we use them instead of code-generating more stuff?
https://www.logicbig.com/tutorials/misc/typescript/generic-type-aliases.html

Reply:
You're right. but the main reason for that is the Tuple type. In Java, it was solved with Tuple2, Tuple3, etc. which I find not so handy. I will change it to Tuple<T extends any[]> to make possible use of Tuple<[string, number]>. it's not so beautiful, but will do the work

Reply:
For tuples, we should use typescript native tuples: https://www.tutorialsteacher.com/typescript/typescript-tuple

Reply:
Yes. This is basically what I suggested. I personally think it's a hack (because it actually uses a list), but it would work.